src/core/createConsent.js

```diff
@@ -0,0 +1,273 @@
+      optOutPreferencesByPurpose = _optOutPreferencesByPurpose;
+      onChange(getStatus());
+    },
+    refreshFromCookies() {
+      refreshFromCookies();
+      onChange(getStatus());
+    },
+    getStatus
+  };
+};
+
+export default ({
+  config,
+  logger,
+  lifecycle,
+  createConsentRequestPayload,
+  sendEdgeNetworkRequest
+}) => {
+  const { orgId, optInEnabled } = config;
+
+  const deferredsAwaitingConsent = [];
+
+  const processConsentStatus = consentStatus => {
+    if (consentStatus.consented === undefined) {
+      return;
+    }
+
+    while (deferredsAwaitingConsent.length) {
+      const deferred = deferredsAwaitingConsent.shift();
+
+      if (consentStatus.consented) {
+        deferred.resolve();
+      } else {
+        deferred.reject(new Error(consentStatus.deniedConsentReason));
+      }
+    }
+  };
+
+  const consentState = createConsentState({
+    optInEnabled,
+    optInCookieName: getNamespacedCookieName(orgId, OPTIN_COOKIE_KEY),
+    optOutCookieName: getNamespacedCookieName(orgId, OPTOUT_COOKIE_KEY),
+    onChange: processConsentStatus
+  });
+
+  const initialConsentStatus = consentState.getStatus();
+
+  if (initialConsentStatus.consented === undefined) {
+    logger.warn("Some commands may be delayed until the user opts in.");
+  } else if (!initialConsentStatus.consented) {
+    logger.warn(
+      `Some commands may be fail. ${initialConsentStatus.deniedConsentReason}`
+    );
+  }
+
+  return {
+    /**
+     * Update the purposes the user has opted into. Only to be called by the
+     * Privacy component.
+     * @param {Object} preferencesByPurpose An object where the key is
+     * the purposes name and the value is a boolean indicating whether the user
+     * has opted into the purpose.
+     */
+    setOptInPurposes(preferencesByPurpose) {
+      const status = consentState.getStatus();
+
+      // The edge restricts modifying opt-in preferences if the user is opted
+      // out, so we'll prevent the attempt.
+      if (
+        !status.consented &&
+        status.deniedConsentReason === OPTED_OUT_OF_ALL_PURPOSES
+      ) {
+        return Promise.reject(new Error(OPTED_OUT_OF_ALL_PURPOSES));
+      }
+
+      consentState.setOptInPreferences(preferencesByPurpose);
+
+      const payload = createConsentRequestPayload();
+      return lifecycle
+        .onBeforeConsentRequest({
+          payload
+        })
+        .then(() => {
+          payload.setPurposes(preferencesByPurpose);
+          return sendEdgeNetworkRequest({
+            payload,
+            action: "opt-in"
+          });
+        })
+        .then(() => {
+          // We optimistically use the preferences the user provided right away,
+          // but if the server doesn't agree and sends back a cookie value that
+          // conflicts, we adopt whatever the server said since that ultimately
+          // is what will be used if the user refreshes the page.
+          consentState.refreshFromCookies();
```

Comment:
Because we're doing this, there is a race condition if the customer updates consent preferences in succession like this:

```
optIn({
  purposes: "all"
});
// At this point, consentState will show that the user has opted into all purposes, because we immediately use the user's preferences before we get a response back from the server.

optIn({
  purposes: "none"
});
// At this point, consentState will show that the user has opted into all purposes, because we immediately use the user's preferences before we get a response back from the server.
```

then let's say the response for the first command comes back from the server. If everything went well, the cookie will be set to `GENERAL=true`. Since we call `consentState.refreshFromCookies()` here, consentState will now show that the user has opted into all purposes until the response comes back from the second command. During this period, Alloy is technically not behaving according to user preferences, since the last thing the customer did was set `purposes: none`.

This is probably a very niche scenario, but one we should probably account for by using some sequencing logic. Thoughts?

Reply:
I get an error message if I don't put {purposes:"all"} in the optOut command. Is it supposed to be all or none? Will opt-out always be all or nothing? Is specifying necessary?

Reply:
Good questions. For now, you must and can only specify `purposes: "all"`. My understanding is that later this may be broken down into more granular purposes. At that point, you'll be able to do `purposes: "all"` still or control purposes more granularly like this:

```
alloy("optOut", {
  purposes: {
    purposeA: true,
    purposeB: false
  }
});
```

It's also possible that later on we'll allow users to, for lack of a better term, un-opt-out. But for now, un-opting-out isn't supported.

Reply:
This speaks to a greater set of race conditions where the server may set a cookie.  This is not just a consent problem.  Perhaps it could be solved in a part of the network object.  You could add a flag to the payload of "willUpdateConsentCookie" and/or "willUpdateIdentityCookie" which would sequence the requests.

Reply:
@jonsnyder let's try to discuss this during today's standup. I want to make sure I understand what you're describing.